docs/DOCUMENTATION_GUIDELINES.md

```diff
@@ -0,0 +1,227 @@
+# Substrate Documentation Guidelines
+
+This document is only focused on documenting parts of substrate that relates to its external API. The list of such crates can be found in [CODEOWNERS](./CODEOWNERS). Search for the crates that are auto-assigned to a team called `docs-audit`.
+
+These are crates that are often used by external developers and deserve a better documents. Consequently, most of this revolves around FRAME-development.
+
+- [Substrate Documentation Guidelines](#substrate-documentation-guidelines)
+  - [General/Non-Pallet Crates](#generalnon-pallet-crates)
+    - [What to Document?](#what-to-document)
+      - [Rust Docs vs. Code Comments](#rust-docs-vs-code-comments)
+    - [How to Document?](#how-to-document)
+    - [TLDR](#tldr)
+    - [Other Guidelines](#other-guidelines)
+      - [Document Through Code](#document-through-code)
+      - [Formatting Matters](#formatting-matters)
+  - [Pallet Crates](#pallet-crates)
+    - [Top Level Pallet Docs (`lib.rs`)](#top-level-pallet-docs-librs)
+    - [Dispatchable](#dispatchable)
+    - [Storage Items](#storage-items)
+
+
+## General/Non-Pallet Crates
+
+First, consider the case for all such crates, except for those that are pallets.
+
+### What to Document?
+
+The first question is, what should you document? Use the following filter:
+
+1. Within the set of crates above,
+2. Any `pub` item within them needs to be documented. If it is not `pub`, it won't show up in the rust-docs, and won't matter.
+    * Within `pub` items, sometimes they are only `pub` in order to be used by another internal crate, and you can foresee that this will not be used by anyone else other than you. These need **not** be documented thoroughly, and are left to your discretion to identify.
+    * Reminder: `trait` items are public by definition, if the trait is public.
+3. Of course, all of these crates should have a reasonable module documentation (`//!`).
+
+
+#### Rust Docs vs. Code Comments
+
+Note that anything starting with `///` is external rust-doc, and anything starting with `//` is just for you to see. Don't mix the two!
+
+```rust
+/// Computes the square root of the input, returning `Ok(_)` if successful.
+///
+/// # Errors
+/// ...
+///
+/// # Complexity
+/// ...
+///
+// Details about the complexity, how you implemented this, and some quirks that
```

Comment:
We created the section `#Complexity` above but are saying that the details about complexity should start with `//`.

Reply:
I think Kian means to note the complexity in the `///` comment, but leave the gritty details of the complexity (why, trade-offs between different considered approaches, etc) to the `//` comment.

Reply:
I meant to say more details of how the implementation is achieved and so on, but I get that this is very confusing 🙈 edited.